biocrnpyler/mixture.py

```diff
@@ -411,12 +414,32 @@ def compile_crn(self) -> ChemicalReactionNetwork:
         #add the extra species to the CRN
         self.add_species_to_crn(self.added_species, component = None)
 
+        #local component enumeration
+        new_comp = []
+        comp_i = 0
+        comps_to_enumerate = self.components
+        #component enumeration
+        outstanding_components = []
+        for a in range(self.recursion_depth):
+            for component in comps_to_enumerate:
+                endpt_components,unenumerated_components = component.update_components()
+                new_comp += endpt_components
+                outstanding_components += unenumerated_components
+            if(len(outstanding_components) == 0):
+                #this happens if we are done early
+                break
+            else:
+                comps_to_enumerate = outstanding_components
+                outstanding_components = []
+        if(len(outstanding_components) > 0):
+            warn("Mixture was left with unenumerated components "+str(', '.join(outstanding_components)))
```

Comment:
Maybe we can make a helper function in mixture, like Mixture.enumerate_components() which does all this logic so the compile loop is cleaner?